File: electron/terminal/ClaudeDetector.ts
export type ClaudeState =
  | "inactive"
  | "idle"
  | "thinking"
  | "generating"
  | "tool_use"
  | "error";

type StateChangeCallback = (state: ClaudeState) => void;

const DEBOUNCE_MS = 100;
const BUFFER_MAX_LINES = 20;

const ANSI_REGEX = /\x1b\[[0-9;]*[a-zA-Z]|\x1b\].*?(\x07|\x1b\\)/g;

const CLAUDE_LAUNCH_REGEX = /(?:^|\n)\s*[$>]\s+claude\b/;
const BARE_PROMPT_REGEX = /(?:^|\n)\s*[$>]\s*$/;

const ERROR_PATTERNS = [/error/i, /failed/i, /exception/i];
const TOOL_USE_PATTERNS = [
  /\btool\b/i,
  /\bread\b.*file/i,
  /\bwrite\b.*file/i,
  /\bedit\b/i,
  /\bbash\b/i,
  /\bgrep\b/i,
  /\bglob\b/i,
];
const THINKING_PATTERNS = [/thinking/i, /\.\.\.\s*$/];
const GENERATING_PATTERNS = [/generating/i, /writing/i];

export class ClaudeDetector {
  private state: ClaudeState = "inactive";
  private isClaudeActive = false;
  private listeners: Set<StateChangeCallback> = new Set();
  private lineBuffer: string[] = [];
  private debounceTimer: ReturnType<typeof setTimeout> | null = null;
  private pendingState: ClaudeState | null = null;

  processOutput(data: string): void {
    const clean = stripAnsi(data);
    const lines = clean.split("\n");

    for (const line of lines) {
      if (line.length > 0) {
        this.lineBuffer.push(line);
      }
    }
    if (this.lineBuffer.length > BUFFER_MAX_LINES) {
      this.lineBuffer = this.lineBuffer.slice(-BUFFER_MAX_LINES);
    }

    const buffered = this.lineBuffer.join("\n");

    if (!this.isClaudeActive) {
      if (CLAUDE_LAUNCH_REGEX.test(buffered)) {
        this.isClaudeActive = true;
        this.setState("idle");
      }
      return;
    }

    if (BARE_PROMPT_REGEX.test(clean)) {
      this.isClaudeActive = false;
      this.setState("inactive");
      return;
    }

    const recentText = this.lineBuffer.slice(-5).join("\n");
    const newState = this.detectState(recentText);
    if (newState !== this.state) {
      this.setState(newState);
    }
  }

  getState(): ClaudeState {
    return this.state;
  }

  onStateChange(callback: StateChangeCallback): () => void {
    this.listeners.add(callback);
    return () => {
      this.listeners.delete(callback);
    };
  }

  private detectState(text: string): ClaudeState {
    if (ERROR_PATTERNS.some((p) => p.test(text))) return "error";
    if (TOOL_USE_PATTERNS.some((p) => p.test(text))) return "tool_use";
    if (THINKING_PATTERNS.some((p) => p.test(text))) return "thinking";
    if (GENERATING_PATTERNS.some((p) => p.test(text))) return "generating";
    return "idle";
  }

  private setState(newState: ClaudeState): void {
    this.pendingState = newState;

    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }

    this.debounceTimer = setTimeout(() => {
      if (this.pendingState !== null && this.pendingState !== this.state) {
        this.state = this.pendingState;
        for (const cb of this.listeners) {
          cb(this.state);
        }
      }
      this.pendingState = null;
      this.debounceTimer = null;
    }, DEBOUNCE_MS);
  }
}

function stripAnsi(str: string): string {
  return str.replace(ANSI_REGEX, "");
}

File: src-tauri/src/terminal/claude_detector.rs
use regex::Regex;
use std::sync::LazyLock;
use std::time::Instant;

const DEBOUNCE_MS: u128 = 100;
const BUFFER_MAX_LINES: usize = 20;

#[derive(Clone, PartialEq, Debug)]
pub enum ClaudeState {
    Inactive,
    Idle,
    Thinking,
    Generating,
    ToolUse,
    Error,
}

impl ClaudeState {
    pub fn as_str(&self) -> &'static str {
        match self {
            ClaudeState::Inactive => "inactive",
            ClaudeState::Idle => "idle",
            ClaudeState::Thinking => "thinking",
            ClaudeState::Generating => "generating",
            ClaudeState::ToolUse => "tool_use",
            ClaudeState::Error => "error",
        }
    }
}

static ANSI_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\x1b\[[0-9;]*[a-zA-Z]|\x1b\].*?(\x07|\x1b\\)").unwrap()
});

static CLAUDE_LAUNCH_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?m)^\s*[$>]\s+claude\b").unwrap()
});

static BARE_PROMPT_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?m)^\s*[$>]\s*$").unwrap()
});

static ERROR_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    vec![
        Regex::new(r"(?i)error").unwrap(),
        Regex::new(r"(?i)failed").unwrap(),
        Regex::new(r"(?i)exception").unwrap(),
    ]
});

static TOOL_USE_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    vec![
        Regex::new(r"(?i)\btool\b").unwrap(),
        Regex::new(r"(?i)\bread\b.*file").unwrap(),
        Regex::new(r"(?i)\bwrite\b.*file").unwrap(),
        Regex::new(r"(?i)\bedit\b").unwrap(),
        Regex::new(r"(?i)\bbash\b").unwrap(),
        Regex::new(r"(?i)\bgrep\b").unwrap(),
        Regex::new(r"(?i)\bglob\b").unwrap(),
    ]
});

static THINKING_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    vec![
        Regex::new(r"(?i)thinking").unwrap(),
        Regex::new(r"\.\.\.\s*$").unwrap(),
    ]
});

static GENERATING_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
    vec![
        Regex::new(r"(?i)generating").unwrap(),
        Regex::new(r"(?i)writing").unwrap(),
    ]
});

pub struct ClaudeDetector {
    state: ClaudeState,
    is_claude_active: bool,
    line_buffer: Vec<String>,
    pending_state: Option<ClaudeState>,
    last_change: Instant,
}

impl ClaudeDetector {
    pub fn new() -> Self {
        Self {
            state: ClaudeState::Inactive,
            is_claude_active: false,
            line_buffer: Vec::new(),
            pending_state: None,
            last_change: Instant::now(),
        }
    }

    /// Process PTY output chunk. Returns Some(new_state) when state changed after debounce.
    pub fn process_output(&mut self, data: &str) -> Option<ClaudeState> {
        let clean = strip_ansi(data);

        for line in clean.split('\n') {
            if !line.is_empty() {
                self.line_buffer.push(line.to_string());
            }
        }
        if self.line_buffer.len() > BUFFER_MAX_LINES {
            let excess = self.line_buffer.len() - BUFFER_MAX_LINES;
            self.line_buffer.drain(..excess);
        }

        let buffered = self.line_buffer.join("\n");

        if !self.is_claude_active {
            if CLAUDE_LAUNCH_RE.is_match(&buffered) {
                self.is_claude_active = true;
                return self.set_state(ClaudeState::Idle);
            }
            return None;
        }

        if BARE_PROMPT_RE.is_match(&clean) {
            self.is_claude_active = false;
            return self.set_state(ClaudeState::Inactive);
        }

        let start = self.line_buffer.len().saturating_sub(5);
        let recent_text = self.line_buffer[start..].join("\n");
        let new_state = detect_state(&recent_text);

        if new_state != self.state {
            return self.set_state(new_state);
        }

        // Check if pending state should be emitted (debounce elapsed)
        self.try_flush_pending()
    }

    pub fn get_state(&self) -> &ClaudeState {
        &self.state
    }

    fn set_state(&mut self, new_state: ClaudeState) -> Option<ClaudeState> {
        self.pending_state = Some(new_state);
        self.last_change = Instant::now();
        self.try_flush_pending()
    }

    fn try_flush_pending(&mut self) -> Option<ClaudeState> {
        if let Some(ref pending) = self.pending_state {
            if self.last_change.elapsed().as_millis() >= DEBOUNCE_MS && *pending != self.state {
                let emitted = pending.clone();
                self.state = emitted.clone();
                self.pending_state = None;
                return Some(emitted);
            }
        }
        None
    }
}

fn detect_state(text: &str) -> ClaudeState {
    if ERROR_PATTERNS.iter().any(|p| p.is_match(text)) {
        return ClaudeState::Error;
    }
    if TOOL_USE_PATTERNS.iter().any(|p| p.is_match(text)) {
        return ClaudeState::ToolUse;
    }
    if THINKING_PATTERNS.iter().any(|p| p.is_match(text)) {
        return ClaudeState::Thinking;
    }
    if GENERATING_PATTERNS.iter().any(|p| p.is_match(text)) {
        return ClaudeState::Generating;
    }
    ClaudeState::Idle
}

fn strip_ansi(s: &str) -> String {
    ANSI_RE.replace_all(s, "").to_string()
}
